src/node_crypto.cc

```diff
@@ -3105,14 +3113,57 @@ void CipherBase::InitIv(const FunctionCallbackInfo<Value>& args) {
   ASSIGN_OR_RETURN_UNWRAP(&cipher, args.Holder());
   Environment* env = cipher->env();
 
-  CHECK_GE(args.Length(), 3);
+  CHECK_GE(args.Length(), 4);
 
   const node::Utf8Value cipher_type(env->isolate(), args[0]);
   ssize_t key_len = Buffer::Length(args[1]);
   const char* key_buf = Buffer::Data(args[1]);
   ssize_t iv_len = Buffer::Length(args[2]);
   const char* iv_buf = Buffer::Data(args[2]);
-  cipher->InitIv(*cipher_type, key_buf, key_len, iv_buf, iv_len);
+  CHECK(args[3]->IsInt32());
+  // Don't assign to cipher->auth_tag_len_ directly; the value might not
+  // represent a valid length at this point.
+  int auth_tag_len = args[3]->Int32Value();
+
+  cipher->InitIv(*cipher_type, key_buf, key_len, iv_buf, iv_len, auth_tag_len);
+}
+
+
+bool CipherBase::InitAuthenticated(const char *cipher_type, int iv_len,
+                                   int auth_tag_len) {
+  // TODO(tniessen) Use EVP_CTRL_AEAD_SET_IVLEN when migrating to OpenSSL 1.1.0
+  static_assert(EVP_CTRL_CCM_SET_IVLEN == EVP_CTRL_GCM_SET_IVLEN,
+                "OpenSSL constants differ between GCM and CCM");
+  if (IsAuthenticatedMode() &&
+      !EVP_CIPHER_CTX_ctrl(ctx_, EVP_CTRL_GCM_SET_IVLEN, iv_len, nullptr)) {
+    EVP_CIPHER_CTX_free(ctx_);
```

Comment:
I would suggest moving the `EVP_CIPHER_CTX_free()` calls to the callers.  For ease of comprehension, clean-up on error should take place in the function that allocated the resource.

You can automate that with `std::unique_ptr`, by the way:
```c++
#include <memory>
// ...
auto d = [this] (...) {
  EVP_CIPHER_CTX_free(ctx_);
  ctx_ = nullptr;
};
std::unique_ptr<void, decltype(d)> free_on_return(&d, d);
// ..
free_on_return.release();  // if the happy path was taken
```